### PR TITLE
Support drop into editor

### PR DESCRIPTION
**Pull request by randomVariable2000**
#63

### CODE REVIEW

File: src/extension.ts
Comment (tahabasri) [inline src/extension.ts:527]: instead of `{ language: '*', scheme: '*'}`, you can simply use `'*'`

Comment (tahabasri) [inline src/extension.ts:527]: you can use anonymous class here, replace `new class DropEditorProvider implements vscode.DocumentDropEditProvider {}` with simple `{}`

File: src/extension.ts
Comment (tahabasri) [inline src/extension.ts:549]: I see a different return expression in the documentation here https://github.com/microsoft/vscode-extension-samples/blob/main/drop-on-document/src/extension.ts#L35, I'm not sure what's the best option. Let's use the best one.

**Review by tahabasri — CHANGES_REQUESTED** [review]
Thanks @randomVariable2000 for the change request. Did you test the code ? I can't seem to make it work in my instance. I keep dropping with no result. I also left some changes I'd see necessary before merging the code.

**Comment by randomVariable2000** [comment]
> Thanks @randomVariable2000 for the change request. Did you test the code ? I can't seem to make it work in my instance. I keep dropping with no result. I also left some changes I'd see necessary before merging the code.

Thanks for reviewing this pull request! I had test the code. It work in my instance.
Can I ask if you held down the 'left shift' key when dropping the snippets?

image

If you did, could you provide some test cases that do not work?

File: src/extension.ts
Comment (randomVariable2000) [reply src/extension.ts]: Done.

Comment (randomVariable2000) [reply src/extension.ts]: Done.

File: src/extension.ts
Comment (randomVariable2000) [reply src/extension.ts]: I noticed that there are several snippet syntaxes available.

If we use draggedSnippet.value as insertText directly, we cannot take advantage of the snippets variable. If we use DocumentDropEdit to appendText or appendVariable, we need to know the type of the snippet, which can be difficult. Therefore, I use the same logic as the open command to insert the snippet by using vscode.commands.executeCommand("editor.action.insertSnippet", xx) or vscode.window.activeTextEditor.edit(edit => { edit.insert(position, xxx); }).

File: src/extension.ts
Comment (tahabasri) [inline src/extension.ts:527]: I refactored your code a bit to reduce amount of if blocks and to use Snippet mimetype instead of text/uri-list. Check comment in Provider file

```
    context.subscriptions.push(vscode.languages.registerDocumentDropEditProvider('*', {
        async provideDocumentDropEdits(
            _document: vscode.TextDocument,
            position: vscode.Position,
            dataTransfer: vscode.DataTransfer,
            _token: vscode.CancellationToken
        ): Promise<vscode.DocumentDropEdit | undefined> {
            const dataItem = await dataTransfer.get('application/vnd.code.tree.snippetsProvider')?.asString();
            if (!dataItem) {
                return;
            }
            try {
                const parsedSource = JSON.parse(dataItem) as readonly Snippet[];
                // only accept one snippet (not a folder)
                if (parsedSource?.length !== 1 || parsedSource[0].folder) {
                    return;
                }
                const draggedSnippet = parsedSource[0];
                // same as open snippet command
                if (draggedSnippet.resolveSyntax === undefined) {
                    draggedSnippet.resolveSyntax = true;
                }
                if (draggedSnippet.resolveSyntax) {
                    vscode.commands.executeCommand("editor.action.insertSnippet", { snippet: draggedSnippet.value });
                } else {
                    const editor = vscode.window.activeTextEditor;
                    if (!editor) {
                        return;
                    }
                    editor.edit(edit => { edit.insert(position, draggedSnippet.value ?? ''); });
                }
            } catch {
                // throws error when parsing `dataItem?.value`, just skip
            }
        }
    }));
```

File: src/provider/snippetsProvider.ts
Comment (tahabasri) [inline src/provider/snippetsProvider.ts:11]: we won't need new mime type, we can use default one `application/vnd.code.tree.snippetsProvider`

File: src/provider/snippetsProvider.ts
Comment (tahabasri) [inline src/provider/snippetsProvider.ts:12]: instead of this one, please use our default `application/vnd.code.tree.snippetsProvider`

File: src/provider/snippetsProvider.ts
Comment (tahabasri) [inline src/provider/snippetsProvider.ts:16]: no need

**Review by tahabasri — CHANGES_REQUESTED** [review]
The new code is working as expected. Thanks for the changes, I did update the code slightly to use the default mimetype, I'll merge it once you apply what I put in comments. Many thanks !

File: src/provider/snippetsProvider.ts
Comment (randomVariable2000) [reply src/provider/snippetsProvider.ts]: Done.

File: src/provider/snippetsProvider.ts
Comment (randomVariable2000) [reply src/provider/snippetsProvider.ts]: Done.

Comment (randomVariable2000) [reply src/provider/snippetsProvider.ts]: Done.

**Review by tahabasri — APPROVED** [review]

**Comment by tahabasri** [comment]
Thanks @randomVariable2000 for your contribution, this will help future users a lot.

**Comment by randomVariable2000** [comment]
Thanks.

**Pull request merged**
